oom: Support for user services

DaanDeMeyer opened the pull request:
Last commit contains more context.

One thing I'm not sure about is if we need to add extra logic to handle reconnects if systemd-oomd restarts. Currently, we reconnect as soon as the user manager has a new ManagedOOM update to send to systemd-oomd. However, if there's no update, systemd-oomd will be without the initial ManagedOOM message for a long time. We could mitigate this by adding a retry timer that tries to reconnect every X seconds when a user manager disconnects from systemd-oomd but I'm not sure if we need to go that far.

Depends on #20658 (not really but it's included for debugging purposes)

bluca commented on src/oom/oomd-manager.c:
```diff
@@ -40,93 +40,98 @@ static int managed_oom_mode(const char *name, JsonVariant *v, JsonDispatchFlags
+                r = -EIO;
+                log_debug("Error getting ManagedOOM cgroups: %s", error_id);
+                goto finish;
+
```
extra new line?

bluca commented on src/oom/oomd-manager.c:
```diff
@@ -40,93 +40,98 @@ static int managed_oom_mode(const char *name, JsonVariant *v, JsonDispatchFlags
+        if (error_id) {
+                r = -EIO;
+                log_debug("Error getting ManagedOOM cgroups: %s", error_id);
+                goto finish;
```
use `else {` and get rid of the goto?

DaanDeMeyer replied:
I'd have to change it back to a goto in the next commit when I add another potentially failing call here so it seems better to keep it as is.

bluca commented on src/core/core-varlink.c:
```diff
@@ -185,6 +172,34 @@ static int vl_method_subscribe_managed_oom_cgroups(
+                JsonVariant *parameters,
+                VarlinkMethodFlags flags,
+                void *userdata) {
+        _cleanup_(json_variant_unrefp) JsonVariant *v = NULL;
```
missing empty new line

bluca commented:
TEST-55-OOMD so might be related

bluca commented on units/systemd-oomd.socket:
```diff
@@ -0,0 +1,21 @@
+Before=sockets.target
+
+[Socket]
+ListenStream=/run/systemd/oom/io.system.ManagedOOM
```
io.systemd?

bluca replied:
oh we use io.system, interesting

DaanDeMeyer commented:
Didn't look at the test yet so I'm assuming it's still gonna fail

DaanDeMeyer commented on src/core/core-varlink.c:
```diff
@@ -469,9 +516,73 @@ int manager_varlink_init(Manager *m) {
+                return log_error_errno(r, "Failed to attach varlink connection to event loop: %m");
+
+        /* Queue the initial ManagedOOM update. */
+        (void) manager_varlink_send_managed_oom_initial(m);
```
When sending ManagedOOM updates, we also ignore the results (grep for usage of manager_varlink_send_managed_oom_update()) so I did the same here but I'm not entirely sure about it. If I understand correctly, we also don't close the connection if sending the initial ManagedOOM update fails from PID1 to oomd so it seemingly makes sense that we do the same for communication between user managers and systemd-oomd.

anitazha replied:
I think I wrote the calls to manager_varlink_send_managed_oom_update() to ignore return since it wasn't critical to the operation in the context of the function. e.g. if we're writing dbus values or processing unit changes. And we would send updates again if any units change. I think it's fine to do the same here given that we'll send unit updates later on as well (or if the user manager has to reconnect to systemd-oomd).

DaanDeMeyer commented:
This excerpt from the oomd logs seems relevant:

```
Sep 10 14:59:49 image systemd-oomd[34]: Memory pressure for /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice is 64.15% > 1.00% for > 30s with reclaim activity
Sep 10 14:59:49 image systemd-oomd[34]: oomd attempting to kill 63 with KILL
Sep 10 14:59:49 image systemd-oomd[34]: oomd attempting to kill 66 with KILL
Sep 10 14:59:49 image systemd-oomd[34]: Killed /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice/testsuite-55-testbloat.service due to memory pressure for /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice be>
Sep 10 14:59:49 image systemd[1]: testsuite-55-testbloat.service: systemd-oomd killed 2 process(es) in this unit.
Sep 10 14:59:49 image systemd[1]: testsuite-55-testbloat.service: Main process exited, code=killed, status=9/KILL
Sep 10 14:59:49 image systemd[1]: testsuite-55-testbloat.service: Failed with result 'signal'.
Sep 10 14:59:49 image systemd[1]: testsuite-55-testbloat.service: Triggering OnFailure= dependencies.
Sep 10 14:59:49 image systemd[1]: testsuite-55-testbloat.service: Consumed 5.688s CPU time.
Sep 10 14:59:49 image systemd-oomd[34]: oomd: New incoming message: {"parameters":{"cgroups":[{"mode":"auto","path":"/testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice/testsuite-55-testbloat.service","property":"ManagedOOM>
Sep 10 14:59:49 image systemd-oomd[34]: oomd: Changing state awaiting-reply-more → processing-reply
Sep 10 14:59:49 image systemd-oomd[34]: oomd: Changing state processing-reply → awaiting-reply-more
Sep 10 14:59:49 image systemd[1]: Starting issue-17602@testsuite-55-testbloat.service.service...
Sep 10 14:59:49 image echo[98]: issue-17602@testsuite-55-testbloat.service.service
Sep 10 14:59:49 image systemd[1]: issue-17602@testsuite-55-testbloat.service.service: Deactivated successfully.
Sep 10 14:59:49 image systemd[1]: Finished issue-17602@testsuite-55-testbloat.service.service.
Sep 10 14:59:49 image systemd-oomd[34]: oomd: New incoming message: {"parameters":{"cgroups":[{"mode":"auto","path":"/system.slice/system-issue\\x2d17602.slice/issue-17602@testsuite-55-testbloat.service.service","property":"ManagedOOMS>
Sep 10 14:59:49 image systemd-oomd[34]: oomd: Changing state awaiting-reply-more → processing-reply
Sep 10 14:59:49 image systemd-oomd[34]: oomd: Changing state processing-reply → awaiting-reply-more
Sep 10 15:00:04 image systemd-oomd[34]: Memory pressure for /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice is 16.05% > 1.00% for > 30s with reclaim activity
Sep 10 15:00:04 image systemd-oomd[34]: oomd attempting to kill 62 with KILL
Sep 10 15:00:04 image systemd-oomd[34]: Killed /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice/testsuite-55-testchill.service due to memory pressure for /testsuite.slice/testsuite-55.slice/testsuite-55-workload.slice be>
Sep 10 15:00:04 image systemd[1]: testsuite-55-testchill.service: systemd-oomd killed 1 process(es) in this unit.
Sep 10 15:00:04 image systemd[1]: testsuite-55-testchill.service: Main process exited, code=killed, status=9/KILL
Sep 10 15:00:04 image systemd[1]: testsuite-55-testchill.service: Failed with result 'signal'.
Sep 10 15:00:04 image systemd[1]: testsuite-55-testchill.service: Triggering OnFailure= dependencies.
```

testchill is also being killed because the memory pressure is still too high after killing testbloat.service. Trying to reproduce this in a mkosi container, I get the same results (testchill is being killed) with and without my changes. If I up the memory pressure limit to 20% I get the expected result in my mkosi container. Are we seeing this test fail elsewhere? I'm slightly unsure if it's caused by my changes or something else.

yuwata commented:
@DaanDeMeyer Maybe #20118.

DaanDeMeyer commented:
Yes looks like the same issue

anitazha commented:
Looking at the journal logs for CentOS CI (Arch Linux) I see `Failed to connect to bus: $DBUS_SESSION_BUS_ADDRESS and $XDG_RUNTIME_DIR not defined (consider using --machine=<user>@.host --user to connect to bus of other user)`. There is a brief period in the extended where memory pressure can remain high and testchill can be killed, but as long as the checks pass before that it is not an issue.

I'll try changing the parameters for  #20118 anyways and see if it helps

bluca commented:
`TEST-55-OOMD` still failing

DaanDeMeyer commented:
@mrc0mmand I found the following log message in the failed sanitizer run logs:

> sep 15 14:44:29 H testsuite-55.sh[421]: ==421==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.

Is that expected or is something wrong with the sanitizers?

mrc0mmand commented:
> @mrc0mmand I found the following log message in the failed sanitizer run logs:
> 
> > sep 15 14:44:29 H testsuite-55.sh[421]: ==421==ASan runtime does not come first in initial library list; you should either link runtime to your application or manually preload it with LD_PRELOAD.
> 
> Is that expected or is something wrong with the sanitizers?

That's expected, unfortunately. I'll drop the oomd test from the sanitizers run (for now) and restart the job. When I have a bit of spare time I'll take a look on how to properly work around it, but it's definitely not the test's fault.

poettering commented on src/oom/oomd-manager.h:
```diff
@@ -3,6 +3,7 @@
 #include "conf-parser.h"
 #include "oomd-util.h"
+#include "sd-bus.h"
 #include "sd-event.h"
```
actually, we generally include `sd-*.h` typically before the rest of our own headers, as kind of a test they actually work with as little else pulled in correctly. i.e. please move these two lines up

poettering commented on src/core/manager.h:
```diff
@@ -448,6 +448,8 @@ struct Manager {
         VarlinkServer *varlink_server;
         /* Only systemd-oomd should be using this to subscribe to changes in ManagedOOM settings */
```
maybe update the comment making clear the this is a varlink connection where we are server and oomd connected to us and that this is used only if we run as system manager

poettering commented on src/core/manager.h:
```diff
@@ -448,6 +448,8 @@ struct Manager {
         VarlinkServer *varlink_server;
         /* Only systemd-oomd should be using this to subscribe to changes in ManagedOOM settings */
         Varlink *managed_oom_varlink_request;
+        /* Used by user managers to send ManagedOOM settings changes to systemd-oomd. */
```
conversely menion here that this is a varlink connection where we are the client, and we connected to oomd, and that we use this only if we run as user manager

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -75,6 +78,21 @@ static int process_managed_oom_message(Manager *m, JsonVariant *parameters) {
+                        * update for a cgroup that doesn't belong to the user. */
+                        if (uid != cg_uid)
+                                return log_error_errno(SYNTHETIC_ERRNO(EPERM),
+                                                       "cgroup path owner UID does not match sender uid (%i != %i)",
```
UID_FMT

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -65,6 +67,7 @@ static int process_managed_oom_message(Manager *m, JsonVariant *parameters) {
                 OomdCGroupContext *ctx;
                 Hashmap *monitor_hm;
                 loadavg_t limit;
+                uid_t cg_uid;
```
move to innermost scope

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -652,6 +696,37 @@ static int manager_connect_bus(Manager *m) {
+        if (r < 0)
+                return log_error_errno(r, "Failed to register varlink method: %m");
+
+        r = varlink_server_listen_fd(s, SD_LISTEN_FDS_START + 0);
```
i'd prefer if we could run this from the cmdline for debug purposes, i.e. in gdb and this would still work. i.e. use the listen fd logic if there is a socket passed, otherwise just bind the socket yourself

poettering commented on src/basic/def.h:
```diff
@@ -64,4 +64,5 @@
         }
 
-#define VARLINK_ADDR_PATH_MANAGED_OOM "/run/systemd/io.system.ManagedOOM"
+#define VARLINK_ADDR_PATH_MANAGED_OOM_SYSTEM "/run/systemd/io.system.ManagedOOM"
```
maybe add comment here, that this is what PID 1 listens on, and where oomd is a client.

poettering commented on src/basic/def.h:
```diff
@@ -64,4 +64,5 @@
 
-#define VARLINK_ADDR_PATH_MANAGED_OOM "/run/systemd/io.system.ManagedOOM"
+#define VARLINK_ADDR_PATH_MANAGED_OOM_SYSTEM "/run/systemd/io.system.ManagedOOM"
+#define VARLINK_ADDR_PATH_MANAGED_OOM_USER "/run/systemd/oom/io.system.ManagedOOM"
```
and conversely, add comment here that this is wat oomd listens on and where the user service managers connect to

poettering commented on src/oom/oomd-manager.h:
```diff
@@ -53,6 +53,7 @@ struct Manager {
         sd_event_source *mem_pressure_context_event_source;
 
         Varlink *varlink;
+        VarlinkServer *varlink_server;
```
as above add comments here, that make clear who listens here and who is client

poettering replied:
maybe even rename the `varlink` field `varlink_client` or so?

poettering commented on src/oom/oomd-manager.c:
```diff
@@ -279,9 +322,9 @@ static int acquire_managed_oom_connect(Manager *m) {
+        r = varlink_connect_address(&link, VARLINK_ADDR_PATH_MANAGED_OOM_SYSTEM);
         if (r < 0)
-                return log_error_errno(r, "Failed to connect to %s: %m", VARLINK_ADDR_PATH_MANAGED_OOM);
+                return log_error_errno(r, "Failed to connect to %s: %m", VARLINK_ADDR_PATH_MANAGED_OOM_SYSTEM);
```
doesn't matter, but you could actually write this as:

```c
return log_error_errno(r, "Failed to connect to " VARLINK_ADDR_PATH_MANAGED_OOM_SYSTEM ": %m");
``` 
and it would be slightly more efficeint

poettering commented on src/core/core-varlink.c:
```diff
@@ -485,9 +516,73 @@ int manager_varlink_init(Manager *m) {
+        if (access(VARLINK_ADDR_PATH_MANAGED_OOM_USER, F_OK) < 0) {
+                log_debug("systemd-oomd varlink unix socket not found, skipping user manager varlink setup");
+                return 0;
+        }
```
please drop this, varlink_connect_address() will return an error if the socket is not connectible, you can just look for thta.

i.e. if varlink_connect_address() returns ENONENT or ERRNO_IS_DISCONNECT() you should assume oomd is not around.

Doing this in two steps has this racy feel to it

poettering commented on src/core/core-varlink.c:
```diff
@@ -120,7 +129,13 @@ int manager_varlink_send_managed_oom_update(Unit *u) {
+                r = varlink_notify(u->manager->managed_oom_varlink_request, v);
+
+        if (u->manager->managed_oom_varlink)
+                r = varlink_send(u->manager->managed_oom_varlink, "io.systemd.oom.SendManagedOOMCGroups", v);
```
the method name is misleading. It suggests that oomd would send something, but it's of course the service manager that sends the change notification here.

maybe call the function ReportManagedOOMCGroupChange() or so?

DaanDeMeyer replied:
Renamed to ReportManagedOOMCGroups to be consistent with SubscribeManagedOOMCGroups in naming

poettering commented on src/core/manager.h:
```diff
@@ -448,6 +448,8 @@ struct Manager {
         /* Only systemd-oomd should be using this to subscribe to changes in ManagedOOM settings */
         Varlink *managed_oom_varlink_request;
+        /* Used by user managers to send ManagedOOM settings changes to systemd-oomd. */
+        Varlink *managed_oom_varlink;
```
maybe even rename these two fields to:

managed_oom_varlink_system and managed_oom_user? 

i.e. i'd really like to underline what these two things are for, and now that they are two this migh tmean that the existing object should be renamed to make things clearer.

poettering reviewed (changes requested):
looks pretty good to me

DaanDeMeyer commented:
Addressed comments and modified the manager logic to use the same varlink object for the subscription and the client connection since we only ever need one depending on whether we're a system or user manager.

poettering commented on src/core/core-varlink.c:
```diff
@@ -93,9 +93,19 @@ int manager_varlink_send_managed_oom_update(Unit *u) {
+                r = manager_varlink_init(u->manager);
+                if (r <= 0)
+                        return r;
+        }
```
maybe merge these conditionalizations? 

```c
if (MANAGER_IS_SYSTEM(u->manager)) {
        /* In system mode we can't send any notifications unless oomd connected back to us. In this mode oomd must initiate communication, not us. */
        if (!u->manager->managed_oom_varlink)
                return 0;
} else {
        /* If we are in user mode, let's connect to oomd if we aren't connected yet. In this mode we must initiate communication to oomd, not the other way round*/
        r = manager_varlink_init(u->manager);
        if (r <= 0)
                return r;        
}
```

poettering commented on src/core/core-varlink.c:
```diff
@@ -120,7 +130,12 @@ int manager_varlink_send_managed_oom_update(Unit *u) {
 
-        return varlink_notify(u->manager->managed_oom_varlink_request, v);
+        if (MANAGER_IS_SYSTEM(u->manager))
+                r = varlink_notify(u->manager->managed_oom_varlink, v);
```
maybe add explanatory comment here:

```c
/* in system mode, oomd is our client, thus send out notifications as replies to the initiating method call from them */
```

poettering commented on src/core/core-varlink.c:
```diff
@@ -120,7 +130,12 @@ int manager_varlink_send_managed_oom_update(Unit *u) {
+        if (MANAGER_IS_SYSTEM(u->manager))
+                r = varlink_notify(u->manager->managed_oom_varlink, v);
+        else
+                r = varlink_send(u->manager->managed_oom_varlink, "io.systemd.oom.ReportManagedOOMCGroups", v);
```
and here a comment like:

```c
/* in user mode, we are oomd's client, thus send out notifications as method callls that do not expect a reply */
```

poettering reviewed (changes requested):
Looks pretty good.  mean, it's a bit weird having the varlink connection in the same field even though sometimes it flows in one dir and sometimes in the other. But i guess it's ok, afte rall we use the same Varlink type for it too, and if it carries enough comments...

anitazha reviewed (commented):
Thanks for working on this it looks great!
